### DISCUSSION TITLE
Feature Request - CLI statement editing

### DISCUSSION THREAD

**Discussion by maurice1408** [duckdb/duckdb#10270] [Ideas]
Make the command line editing easier by either

1. Support readline or vim style keyboard shortcuts e.g. Ctrl-A (readline) start of line or 'b' go back one word (vim) etc
2. or... allow invocation of external editor ($EDITOR) to edit current command line

Both snowsql and the MySQL clients implement one or both of the above.

**Comment by jgrg** [comment]
Just found this when I was about to start my own discussion:

Database clients, such as PostgreSQL's `psql`, have an option to open the previous statement in the history in whatever editor is specified in the `EDITOR` environment variable. In `psql` you type `\e`. I find this really useful when developing a complex query, and I'd love to have this in duckdb too.

An additional feature I implemented in a Perl SQL client I wrote was to format and indent the SQL statement that was sent to the editor, capitalising the SQL keywords, adding line breaks before commas, _etc…_

**Comment by kamatsuoka** [comment]
Came here searching for alt-d (forward delete word) and alt-backspace (backward delete word) support, which would make the cli tool much more usable for me.

But @maurice1408 do other readline-style shortcuts not work for you? For me (v0.9.2, mac os) ctrl-a, ctrl-e, ctrl-k, alt-b, and alt-f all work normally.

**Reply by kamatsuoka** [reply]
Looking into this a little more, it looks like alt-f and alt-b support were added [back in 2021](https://github.com/duckdb/duckdb/issues/1585) 

Following the link to linenoise in that issue, I discovered that the shortcut for deleting the previous word is [ctrl-w](https://github.com/antirez/linenoise/blob/93b2db9bd4968f76148dd62cdadf050ed50b84b3/linenoise.c#L1081). Would be nice if that were configurable!

**Reply by Mytherin** [reply]
I've opened a PR to extend support for [various ALT commands](https://github.com/duckdb/duckdb/pull/10493).

**Reply by maurice1408** [reply]
@kamatsuoka -  thanks indeed some of the readline-style shortcuts do work. My duckdb (0.9.2) build is on Debian bullseye. Would really like the option to invoke whatever editor is in $EDITOR though.

**Comment by Mytherin** [comment]
We've done some work on improving read-line compatibility, happy to extend it further. Alternatively, you can directly use read-line using `rlwrap`:

```bash
rlwrap --substitute-prompt="> " duckdb -batch
```

**Reply by vlulla** [reply]
While read-line capabilities are helpful, i think having something like psql's `\e`, that @jgrg mentions, is quite useful. For instance, I use this query (which uses digraphs[^digraph-benefit] from vim) for learning about duckdb:

```sql
select function_name,function_type,
list_transform(list_zip(parameters,parameter_types),x->x[1]||' «'||x[2]||'»') as params,return_type,
example from duckdb_functions() where schema_name='main' order by random() limit 5;
```

Typing this query is not straightforward on the cli. How i input this query now is I have a tmux pane with vim open where i type the query and then copy/paste the query into duckdb cli. This minor friction would be reduced tremendously, making the user experience more pleasant _i hope_, if there was a meta command (maybe named `.edit`?) that opens the editor with the last query. And just fyi, Julia repl also has something like this called `edit` function which is surprisingly useful, especially for exploratory work.

Regardless, thanks for the duckdb cli! It is a joy to work with day in and day out!

[^digraph-benefit]: Digraphs, or the underlying utf-8 code, makes searching for important stuff considerably easier than using regex! I use vim digraphs quite often.

**Reply by Mytherin** [reply]
Thanks - I agree this is also useful to have. I have opened a PR that adds this functionality - see https://github.com/duckdb/duckdb/pull/11447. Perhaps you could give it a shot to see if it works as you expect it to work?

Also CC @jgrg @kamatsuoka

**Reply by vlulla** [reply]
@Mytherin thanks for the prompt work on this feature! I tried this feature and made a small comment, based on my cursory experience, on the PR page. I don't know if you get notifications from the PR page in which case please pardon this redundant notification.

**Reply by jgrg** [reply]
Sorry, didn't see your pull request. I just found that `\e` works by accident in v1.0.0!
I'd like to request that the temporary file is given a `\.sql` extension so that my editor knows it's editing SQL and can do syntax highlighting.

**Reply by Mytherin** [reply]
Sure, I've opened a PR for this here - https://github.com/duckdb/duckdb/pull/12723

**Reply by maurice1408** [reply]
I just came back to look at this issue. @Mytherin thanks for the implementation of the `\e`  !!

Only one comment on `\e`, most clients snowflake, fish command editing etc place the command edited in $EDITOR back at the command line but do not automatically / immediately execute the command - thus giving the user one last chance to review the edited command before hitting enter.  I think this would be the preferred mode but hey grateful for `\e`!

I mirror @vlulla 's comments about the CLI, a joy to work with.

**Reply by vlulla** [reply]
@maurice1408 just for your information,`.edit` used to behave just like how you wish! And, i [requested](https://github.com/duckdb/duckdb/pull/11447#issuecomment-2029732663) to modify the behavior so that it became similar to psql's `\e` meta-command. I was motivated by my familiarity, and fondness of, psql's `\e` command. Had i been aware of the different behavior in other clients that you mentioned i would have not made this suggestion. I was just thrilled to get this `.edit`/`\e` meta-command _in_ duckdb! And, i'll continue to be thrilled using `.edit` regularly even if the behavior changes again.

**Reply by maurice1408** [reply]
@vlulla I saw that in the history, I can live with it, just glad to have the ability to edit in vim (yes vim!)

**Reply by jgrg** [reply]
I _think_ the behaviour is that it does immediately run the command if it ends with a semi-colon. Leave off the trailing semi-colon to see it in the terminal, then type `;<Enter>` to run it.

**Reply by maurice1408** [reply]
@jgrg - good find! What it is actually doing is "waiting" for the semi-colon to indicate end of statement input, just like the CLI does. Good way to work around the immediate execution though. Thanks
